Update realm initialization as a Status List Issuer to align with new JWK format for the public key

Ogenbertrand opened the pull request:
Closes: https://github.com/ADORSYS-GIS/token-status-link/issues/34

mbunwe-victor reviewed (changes requested):
I have added a single comment for you to check, @Ogenbertrand.

Ogenbertrand commented on src/main/java/com/adorsys/keycloakstatuslist/service/StatusListService.java:
```diff
@@ -162,7 +162,7 @@ private void validateStatusRecord(TokenStatusRecord statusRecord) throws StatusL
 
         // Require public_key to be set by the caller (e.g.,
         // TokenStatusEventListenerProvider)
```
DONE @mbunwe-victor !

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/model/TokenStatusRecord.java:
```diff
@@ -160,7 +160,7 @@ public String toString() {
                 ", issuer='" + issuer + '\'' +
-                ", publicKey='" + publicKey + '\'' +
+                ", publicKey=" + publicKey +
                 ", alg='" + alg + '\'' +
```
Convert to a string so it is properly displayed. Or better, use Jackson's writeValueAsString for this entire "toString" method.

IngridPuppet commented on docker-compose.yml:
```diff
@@ -12,10 +12,8 @@ services:
-      --db-url=jdbc:postgresql://postgres:5432/keycloak_token_status_plugin
+    entrypoint: /bin/sh
+    command: -c "/opt/keycloak/bin/kc.sh build && exec /opt/keycloak/bin/kc.sh start-dev --features=oid4vc-vci --log-level=INFO,com.adorsys.keycloakstatuslist:DEBUG --db=postgres --db-username=admin --db-password=admin --db-url=jdbc:postgresql://postgres:5432/keycloak_token_status_plugin"
     environment:
```
What does this update solve? At least there should be a way to have a multiline version of it for readability.

IngridPuppet replied:
It was certainly due to something else, because there is no need for explicit building on these Docker images. Please double check and report the error.

Ogenbertrand replied:
Yes you are correct i reverted to the previous state before doing my last commit it seem the change effected, but for sure it should look goo now. The wasn't any error actually, while starting the compose file i expected the plugin loaded into keycloak after starting keycloak which wasn't the case so i though perhaps it's cause i wasn't building the docker image. but you clarified me previously that we'll load the plugin in the SSI instead.

Ogenbertrand commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceProviderFactory.java:
```diff
@@ -207,17 +208,13 @@ private boolean registerRealmAsIssuer(KeycloakSession session, RealmModel realm)
-                            base64.replaceAll("(.{64})", "$1\n") +
-                            "\n-----END PUBLIC KEY-----";
+                    publicKey = JWKBuilder.create().kid(activeKey.getKid()).algorithm(activeKey.getAlgorithmOrDefault()).rsa(activeKey.getPublicKey());
                 } catch (Exception e) {
```
Should we rely only on the rsa key or confifure it to use diff keys from keycloak @IngridPuppet ?

Ogenbertrand commented:
I have extended this PR to fix an error i encountered while testing, this was what happened the Keycloak container failed to start because of an invalid JPA mapping in StatusListMappingEntity class where @GeneratedValue was incorrectly placed on a non-primary key field (idx), so I refactored the StatusListMappingEntity to make idx the primary key and updated the database schema accordingly. Please check that it makes sense or not.

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/service/RevocationRecordService.java:
```diff
@@ -38,77 +95,37 @@ public TokenStatusRecord createRevocationRecord(CredentialRevocationRequest requ
             logger.errorf("Failed to create revocation record. RequestId: %s, Error: %s", 
                          requestId, e.getMessage());
+            if (e instanceof StatusListException) throw (StatusListException) e;
             throw new StatusListException("Failed to create revocation record: " + e.getMessage(), e);
```
There is no issue wrapping a StatusListException inside another one.
```suggestion
throw new StatusListException("Failed to create revocation record: " + e.getMessage(), e);
```

Hermann-Core commented on src/main/java/com/adorsys/keycloakstatuslist/model/TokenStatusRecord.java:
```diff
@@ -16,7 +27,7 @@ public class TokenStatusRecord {
+    private JWK publicKey;
 
     @JsonProperty("alg")
     private String alg;
```
Is this field still relevant, given that we are removing the `alg` field?

Hermann-Core commented on src/main/java/com/adorsys/keycloakstatuslist/service/RevocationRecordService.java:
```diff
@@ -27,9 +29,64 @@ public RevocationRecordService(KeycloakSession session) {
+                throw new StatusListException("Unsupported key type for realm " + realm.getName() + ": " + pubKey.getClass().getName());
+            }
+            
+            String finalAlg = activeKey.getAlgorithm() != null ? activeKey.getAlgorithm() : algorithm;
```
Isn’t this a potential logical flaw?  
This doesn’t validate compatibility in cases where the `activeKey` doesn’t contain the algorithm and there is a mismatch between the fallback algorithm and the key’s actual algorithm.

Hermann-Core commented on src/main/java/com/adorsys/keycloakstatuslist/service/StatusListService.java:
```diff
@@ -15,6 +15,7 @@
 import org.apache.hc.core5.http.io.entity.EntityUtils;
 import org.apache.hc.core5.http.io.entity.StringEntity;
 import org.jboss.logging.Logger;
+import org.keycloak.jose.jwk.JWK; // Import added
```
Is this comment relevant?
```suggestion
import org.keycloak.jose.jwk.JWK;
```

Hermann-Core commented on src/main/java/com/adorsys/keycloakstatuslist/service/StatusListService.java:
```diff
@@ -137,15 +138,16 @@ public void updateRecord(TokenStatusRecord statusRecord) throws StatusListExcept
         TokenStatusRecord issuerRecord = new TokenStatusRecord();
         issuerRecord.setIssuer(issuerId);
-        issuerRecord.setPublicKey(publicKey);
+        issuerRecord.setPublicKey(publicKey); // Now matches the TokenStatusRecord signature
```
Is the comment here relevant?

Hermann-Core commented on src/main/java/com/adorsys/keycloakstatuslist/service/StatusListService.java:
```diff
@@ -137,15 +138,16 @@ public void updateRecord(TokenStatusRecord statusRecord) throws StatusListExcept
         issuerRecord.setIssuer(issuerId);
-        issuerRecord.setPublicKey(publicKey);
+        issuerRecord.setPublicKey(publicKey); // Now matches the TokenStatusRecord signature
         issuerRecord.setAlg(algorithm);
```
Is the `algorithm` field still relevant in the request body?

Hermann-Core replied:
If I recall correctly, we switched to using the JWK format specifically to avoid having a separate `alg` field in the payload. Given that, I don’t think the `alg` field is required anymore.

Hermann-Core reviewed (commented):
Hello @Ogenbertrand, I’ve added some review comments, could you check and address them?

Ogenbertrand commented on src/main/java/com/adorsys/keycloakstatuslist/service/RevocationRecordService.java:
```diff
@@ -27,9 +29,64 @@ public RevocationRecordService(KeycloakSession session) {
+            
+            if (activeKey == null || activeKey.getPublicKey() == null) {
+                activeKey = keyManager.getActiveKey(realm, KeyUse.SIG, Algorithm.RS256);
+                algorithm = Algorithm.RS256;
```
We don't solely depend on RSA type only it it supports both RSA and EC keys. The plugin automatically uses the server's preferred signing algorithm, like the more efficient EC-256, but defaults to the widely-supported RSA-256 to ensure it works correctly with any standard setup.

Hermann-Core replied:
EC-256 is also widely supported and is generally the recommended choice between the two. Unless you have a strong requirement to use RSA-256, I suggest defaulting to it here.

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/jpa/entity/StatusListMappingEntity.java:
```diff
@@ -16,9 +16,7 @@ public class StatusListMappingEntity {
-    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "status_list_seq_gen")
-    @SequenceGenerator(name = "status_list_seq_gen", sequenceName = "status_list_counter_seq", allocationSize = 1)
+    @Column(name = "idx", nullable = false, updatable = false, insertable = false, columnDefinition = "BIGSERIAL")
     private Long idx;
```
Please revert.

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/jpa/entity/StatusListMappingEntity.java:
```diff
@@ -87,13 +76,12 @@ public void setRealmId(String realmId) {
-        int result = Objects.hashCode(getId());
-        result = 31 * result + Objects.hashCode(getIdx());
+        int result = Objects.hashCode(getIdx());
         result = 31 * result + Objects.hashCode(getStatusListId());
```
Please revert.

IngridPuppet commented on docker-compose.yml:
```diff
@@ -27,4 +27,4 @@ services:
 
 volumes:
-  db_data:
+  db_data:
```
Please revert the final empty line.

IngridPuppet replied:
My comment is about adding one.

Ogenbertrand replied:
fix

IngridPuppet commented on src/test/java/com/adorsys/keycloakstatuslist/StatusListProtocolMapperTest.java:
```diff
@@ -150,7 +150,6 @@ private long mockEntityPersist() {
             StatusListMappingEntity entity = invocation.getArgument(0);
             entity.setIdx(simulatedIndex); // Simulate sequence generation
-            entity.setId(UUID.randomUUID().toString()); // Simulate UUID generation
             return null;
```
Please revert the setId.
